### ISSUE TITLE
Are global or name-agnostic sink and default logger possible?

### ISSUE THREAD

**Issue by Sarcasm** [gabime/spdlog#523]
Sorry for the bad title.

I'm interested in a logging library that I can use in my libraries.
In terms of usage, I would expect the logging library to allow me to retrieve a logger, whether or not the main binary using the library creates an appropriate logger.

Right now, I'm able to achieve this by using a null sink when `spdlog::get()` returns `nullptr`:

*mylib.cpp*:
```cpp
void mylibfunc() {
        auto logger = spdlog::get("mylib");
        if (logger == nullptr) {
            logger = spdlog::create<spdlog::sinks::null_sink_st>("mylib");
        }

       std::error_code ec;
       doSomethingWithError(ec);
       if (ec) {
          logger->warn("warning: failed to do something: {}", ec.message());
          return;
       }
}
```

And in my main binary I would like to be able to create a "catch-all" logger,
so that if I use multiple libraries using `spdlog`, I can create one sink for all of them,
new or existing libraries can introduce new logger, I don't have to update my main binary.
However, I still find useful to customize log level per-logger if the user specifies it to my binary.
So I don't want to know all the possible loggers, but still log them. And I want to be able to tweak known ones.

I'm not sure I can achieve this part at this time, what do you think?

I have in mind something like this, argument-less constructor would create a catch-all sink:

```cpp
int main() {
        auto console = spdlog::stdout_logger_mt();
        ... // call libraries, all logs go to stdout
}
```

Or alternatively, a regexp/globbing could be allowed to all for filtering logger names:

```cpp
int main() {
        auto console = spdlog::stdout_logger_mt("*"); // all loggers go here
        auto libALogger = spdlog::simple_file_sink("libA", "libA.txt"); // libA logger goes here too
        auto libBLogger = spdlog::simple_file_sink("libB|libB.*", "libA.txt"); // libB and any sublogger such as libB.foo, libB.bar go here in addition to the global console logger
        ... // call libraries, logs go in appropriate sinks
}
```

I have this in mind because of a library I was using in my old company, which was very useful usage (https://github.com/Parrot-Developers/ulog/blob/master/libulog/include/ulog.h#L43).
Also I think this is similar to what the Python `logging` module does.

Basically I want libraries to be able to use logging without caring about the sinks.
And I want the binaries to define the sink and not care about the library logger names.
But still, I want the binaries to be able to tweak the log level for some logger names (regex/globbing is not a necessity for me but I thought it was a way to implement it).

**Comment by gabime** [comment]
I am not sure i understand - Is it a global sink with different loggers ,all using the same sink?

**Comment by Sarcasm** [comment]
I guess so, in my specific cases I only want one sink per application, but I want my application to use multiple libraries, each defining their own loggers.

**Comment by gabime** [comment]
So yes. Use a global sink. Pass it around in thr eay you see fit (or get it from the logger->sinks() method)

**Comment by Sarcasm** [comment]
I think I'm missing something, I'm not sure to understand what is a global sink.
Right now I only know how to make specific sink, e.g.:

        auto console = spdlog::stdout_logger_mt("libA");

This sink won't cut it for "libB"?

**Comment by gabime** [comment]
You can create loggers objects and pass sink to the ctor. please refer to the docs for details.

**Comment by Sarcasm** [comment]
Ok, but I think this solution is different from the behavior I described in my first post:

> I want libraries to be able to use logging without caring about the sinks.
> And I want the binaries to define the sink and not care about the library logger names.

Anyway, I think I'm sure this exact behavior is not possible right now, I won't bother you any longer with this, the library looks great otherwise.

Thank you.

**Comment by patrikhuber** [comment]
My use case is actually quite similar, and I was wondering how to achieve that (if it's possible at all). I also think this must be a quite common use case, but I couldn't really find much information about it at all (neither in the docs here nor by googling).
I think @Sarcasm wants/wanted to achieve the same as me, and the important thing is basically:

> * I want libraries to be able to use logging without caring about the sinks.
> * And I want the binaries to define the sink and not care about the library logger names.
> * But still, I want the binaries to be able to tweak the log level for some logger names

So you do the logging inside a library, "sink-agnostic" (the sink is not specified at this point, and the library is compiled on its own.
You then have an executable which may or may not define a sink of any type. If it defines a sink, then all logging calls within the (separately compiled library) are being logged into that sink.

**Comment by gabime** [comment]
Maybe another way to approach this is to have a root logger created in main, and then cloned by the library (as discussed in #776). So  libs can have their own logger instances, but with same sinks and formats as defined in main root logger.

 I will be happy to discuss what can be added to spdlog to achieve this conveniently (add “clone()” function ? another approach?) - seems like a popular demand these days, as more and more libs are using spdlog.

**Comment by russkel** [comment]
Was there any development on this feature request? I would love to use spdlog in libraries if it meant I could redirect the logger output to, for example, ROS2's built in logging functionality.

**Comment by gabime** [comment]
The current possible approach is to pass the libraries a logger instance, which they can clone() and use.

**Comment by russkel** [comment]
Thanks @gabime
